Add GPIO and Hardware Button support

halal-beef opened the pull request:
As the title suggests, this PR adds GPIO and Hardware Button support, sadly in terms of GPIO, only parts of ALIVE are supported right now as that is a direct dependency for the hardware buttons. Some files may have some hacks I don't know how to circumvent, so if you do have any issues and can help, please do :).

halal-beef committed:
fireplace: Add initial Exynos GPIO support

This commit adds support for Exynos GPIO banks,this
supports controlling GPIOs and getting values of GPIOs.
Currently supported GPIO banks are banks that are
located in ALIVE, as they are a direct dependency for
button control.

Signed-off-by: Umer Uddin <umer.uddin@mentallysanemainliners.org>

BotchedRPR commented on emulator/soc/CMakeLists.txt:
```diff
@@ -4,6 +4,6 @@
 pkg_check_modules(UNICORN REQUIRED unicorn)
 
-add_library(fireplace_soc STATIC fb/fb.c memmap.c soc.c uart/uart.c usb/usb.c)
+add_library(fireplace_soc STATIC fb/fb.c memmap.c soc.c uart/uart.c usb/usb.c gpio/exynos_gpio.c gpio/gpio_alive.c)
```
Sort alphabetically

BotchedRPR commented on emulator/soc/gpio/gpio_alive.c:
```diff
@@ -0,0 +1,55 @@
+	exynos_gpio_cfg_pin(uc_s, bank_volume, 4, GPIO_INPUT);
+	exynos_gpio_cfg_pin(uc_s, bank_power, 4, GPIO_INPUT);
+
+	printf("= intialised button pins!\n");
```
typo.

halal-beef replied:
what a late night + the british spelling does to one

BotchedRPR commented on emulator/soc/gpio/gpio_alive.c:
```diff
@@ -0,0 +1,55 @@
+void gpio_alive_hook(uc_engine *uc, uc_mem_type type, uint64_t address, int size, int64_t value, void *user_data)
+{
+	pthread_mutex_lock(&gpio_alive_lock);
+	pthread_mutex_unlock(&gpio_alive_lock);
```
why is this here? it's not doing any operations on the variable

halal-beef replied:
used UART as a base, will remove

BotchedRPR commented on gui/gui_core.c:
```diff
@@ -134,6 +134,23 @@ void blit_uart_window(struct nk_context *ctx)
+	{
+		nk_layout_row_static(ctx, 40, 100, 3);
+
+		if (nk_button_label(ctx, "Power")) {}
```
so these are unused? how is this supposed to work?

halal-beef replied:
unused in first commit, already mentioned
used in the second commit

BotchedRPR replied:
Then I'd add everything for gui in the second commit to avoid dead code

BotchedRPR commented on emulator/core/emulator.c:
```diff
@@ -67,6 +67,7 @@ load_image(uc_engine *uc, const char *file, uint64_t base, uint64_t *last)
 }
 
 atomic_int sharedState = 0;
+struct uc_struct *uc_engine_struct;
```
no! don't do this.

keep unicorn stuff in soc/ and gui stuff in gui/

you should make an extern struct that will handle the button states, that you will access on the gpio hook in soc/

BotchedRPR commented on emulator/soc/CMakeLists.txt:
```diff
@@ -4,6 +4,6 @@
 pkg_check_modules(UNICORN REQUIRED unicorn)
 
-add_library(fireplace_soc STATIC fb/fb.c memmap.c soc.c uart/uart.c usb/usb.c gpio/exynos_gpio.c gpio/gpio_alive.c)
+add_library(fireplace_soc STATIC fb/fb.c memmap.c soc.c uart/uart.c usb/usb.c gpio/exynos_gpio.c gpio/gpio_alive.c hardware_buttons/hardware_buttons.c)
```
sort alphabetically.

BotchedRPR commented on include/fireplace/gui/gui.h:
```diff
@@ -25,9 +25,10 @@ extern pthread_mutex_t main_mutex;
 extern pthread_cond_t main_cond;
 
 extern atomic_int sharedState;
+extern struct uc_struct *uc_engine_struct;
```
ditto comment 1

halal-beef committed:
fireplace: Add hardware button control support

This commit adds hardware button control support,
the buttons targetted are: Power, Volume Up and
Volume Down. This works in conjunction with the
Exynos GPIO support commit.

Signed-off-by: Umer Uddin <umer.uddin@mentallysanemainliners.org>

halal-beef commented:
implemented all changes on https://github.com/halal-beef/fireplace/commit/4043ab2d63654b712d2d7371e1d547006295958b except now it doesn't work

halal-beef commented:
works now LOL

BotchedRPR commented on gui/gui_core.c:
```diff
@@ -134,6 +135,39 @@ void blit_uart_window(struct nk_context *ctx)
+
+	emuState = atomic_load(&sharedState);
+
+	if(nk_begin(ctx, "Hardware Button Control", nk_rect(25, 25, 321, 83),
```
How about only beginning this window when the emulator is started?

halal-beef replied:
sure

halal-beef replied:
Done

BotchedRPR commented:
I'd move emuState to the per frame loop just to not load it twice, but that's a nit I'll get to tomorrow 

Thank you so much Umer!

halal-beef commented:
Thanks for the merge
